### PR TITLE
feat: support update tags

### PR DESCRIPTION
**Pull request by broccoliSpicy**
This PR tries to add update tags functionality to lance, this can allow user to point a existing tag to a different version of the dataset, addresses the deficiency of `delete then re-create` described by #2742.

### CODE REVIEW

**Comment by broccoliSpicy** [comment]
@wjones127 upon working on this issue, I feel like we might be able to utilize some very fancy time travel data structures and computational models to achieve some stronger persistency([full persistency](https://en.wikipedia.org/wiki/Persistent_data_structure), [confluent persistency](https://arxiv.org/abs/1104.3045)), might be a interesting topic to look into.

**Comment by wjones127** [comment]
> @wjones127 upon working on this issue, I feel like we might be able to utilize some very fancy time travel data structures and computational models to achieve some stronger persistency([full persistency](https://en.wikipedia.org/wiki/Persistent_data_structure), [confluent persistency](https://arxiv.org/abs/1104.3045)), might be a interesting topic to look into.

Maybe. I think given transactional mechanics and how our conflict resolution work, it may not be desirable to go far down that path. I don't think "branches" will really be possible, for example.

**Review by westonpace — APPROVED** [review]
Some minor suggestions but looks good to me.  Thanks!

File: python/python/tests/test_dataset.py [hunk 1]
```diff
@@ -290,6 +290,21 @@ def test_tag(tmp_path: Path):
     with pytest.raises(ValueError):
         lance.dataset(base_dir, "missing-tag")
 
+    # test tag update
+    with pytest.raises(ValueError):
+        ds.tags.update("tag1", 3)
+
+    with pytest.raises(ValueError):
+        ds.tags.update("tag3", 1)
```
[the remark this answers is not in the document]
Comment (broccoliSpicy) [reply python/python/tests/test_dataset.py]: This is a good idea, thanks, done.

File: python/src/dataset.rs [hunk 2]
```diff
@@ -1023,6 +1023,19 @@ impl Dataset {
...
 
+    fn update_tag(&mut self, tag: String, version: u64) -> PyResult<()> {
+        let mut new_self = self.ds.as_ref().clone();
+        RT.block_on(None, new_self.tags.update(tag.as_str(), version))?
+            .map_err(|err| match err {
+                lance::Error::NotFound { .. } => PyValueError::new_err(err.to_string()),
+                lance::Error::RefNotFound { .. } => PyValueError::new_err(err.to_string()),
+                lance::Error::VersionNotFound { .. } => PyValueError::new_err(err.to_string()),
+                _ => PyIOError::new_err(err.to_string()),
+            })?;
```
[the remark this answers is not in the document]
Comment (broccoliSpicy) [reply python/src/dataset.rs]: I didn't know we have a python `infer_error`, thanks, moved tag errors into `infer_error`.

**Review by wjones127 — APPROVED** [review]

**Pull request merged**